junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/EnumArgumentsProvider.java

```diff
@@ -26,28 +27,47 @@
  */
 class EnumArgumentsProvider implements ArgumentsProvider, AnnotationConsumer<EnumSource> {
 
-	private EnumSet<?> constants;
+	private EnumSource enumSource;
 
 	@Override
-	@SuppressWarnings({ "unchecked", "rawtypes" })
 	public void accept(EnumSource enumSource) {
-		Class enumClass = enumSource.value();
-		this.constants = EnumSet.allOf(enumClass);
+		this.enumSource = enumSource;
+	}
 
+	@Override
+	public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+		Set<? extends Enum<?>> constants = getEnumConstants(context);
 		EnumSource.Mode mode = enumSource.mode();
 		String[] declaredConstantNames = enumSource.names();
 		if (declaredConstantNames.length > 0) {
 			Set<String> uniqueNames = stream(declaredConstantNames).collect(toSet());
 			Preconditions.condition(uniqueNames.size() == declaredConstantNames.length,
 				() -> "Duplicate enum constant name(s) found in " + enumSource);
-			mode.validate(enumSource, uniqueNames);
-			this.constants.removeIf(constant -> !mode.select(constant, uniqueNames));
+			mode.validate(enumSource, constants, uniqueNames);
+			constants.removeIf(constant -> !mode.select(constant, uniqueNames));
 		}
+		return constants.stream().map(Arguments::of);
 	}
 
-	@Override
-	public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
-		return constants.stream().map(Arguments::of);
+	private <E extends Enum<E>> Set<? extends E> getEnumConstants(ExtensionContext context) {
+		Class<E> enumClass = determineEnumClass(context);
+		return EnumSet.allOf(enumClass);
+	}
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	private <E extends Enum<E>> Class<E> determineEnumClass(ExtensionContext context) {
+		Class enumClass = enumSource.value();
+		if (enumClass.equals(NullEnum.class)) {
```

Comment:
`==` should suffice here -- right?

Reply:
It should unless there are different `ClassLoader` involved. Thus, I usually tend to use `equals()` to be safe.

Reply:
> It should unless there are different `ClassLoader` involved.

Well, `java.lang.Class<T>` does not override `equals()`. So having a different `ClassLoader` would not make a difference in the outcome.

I suppose it's just a matter of taste.